src/fdc3/dotnet/DesktopAgent/src/DesktopAgent/Infrastructure/Internal/Fdc3DesktopAgentMessageRouterService.cs

```diff
@@ -0,0 +1,250 @@
+        }
+
+        if (request.TargetFdc3InstanceId == null || request.Intent == null)
+        {
+            return MessageBuffer.Factory.CreateJson(StoreIntentResultResponse.Failure(ResolveError.IntentDeliveryFailed), _jsonSerializerOptions);
+        }
+
+        var response = await _desktopAgent.StoreIntentResult(request);
+        return MessageBuffer.Factory.CreateJson(response, _jsonSerializerOptions);
+    }
+
+    internal async ValueTask<MessageBuffer?> HandleGetIntentResult(string endpoint, MessageBuffer? payload, MessageContext context)
+    {
+        var request = payload?.ReadJson<GetIntentResultRequest>(_jsonSerializerOptions);
+        if (request == null)
+        {
+            return MessageBuffer.Factory.CreateJson(GetIntentResultResponse.Failure(ResolveError.IntentDeliveryFailed), _jsonSerializerOptions);
+        }
+
+        if (request.TargetAppIdentifier?.InstanceId == null || request.Intent == null || request.MessageId == null)
+        {
+            return MessageBuffer.Factory.CreateJson(GetIntentResultResponse.Failure(ResolveError.IntentDeliveryFailed), _jsonSerializerOptions);
+        }
+
+        var response = await _desktopAgent.GetIntentResult(request);
+        return MessageBuffer.Factory.CreateJson(response, _jsonSerializerOptions);
+    }
+
+    private async ValueTask DisposeSafeAsync(IEnumerable<ValueTask> tasks)
+    {
+        foreach (var task in tasks)
+        {
+            try
+            {
+                await task;
+            }
+            catch (Exception exception)
+            {
+                _logger.LogWarning($"Could not dispose task: {task.IsCompleted}. Exception: {exception}");
+            }
+        }
+    }
+
+    public async Task StartAsync(CancellationToken cancellationToken)
+    {
+        await _messageRouter.RegisterServiceAsync(Fdc3Topic.FindChannel, HandleFindChannel, cancellationToken: cancellationToken);
+        await _messageRouter.RegisterServiceAsync(Fdc3Topic.FindIntent, HandleFindIntent, cancellationToken: cancellationToken);
+        await _messageRouter.RegisterServiceAsync(Fdc3Topic.FindIntentsByContext, HandleFindIntentsByContext, cancellationToken: cancellationToken);
+        await _messageRouter.RegisterServiceAsync(Fdc3Topic.RaiseIntent, HandleRaiseIntent, cancellationToken: cancellationToken);
+        await _messageRouter.RegisterServiceAsync(Fdc3Topic.GetIntentResult, HandleGetIntentResult, cancellationToken: cancellationToken);
+        await _messageRouter.RegisterServiceAsync(Fdc3Topic.SendIntentResult, HandleStoreIntentResult, cancellationToken: cancellationToken);
+        await _messageRouter.RegisterServiceAsync(Fdc3Topic.AddIntentListener, HandleAddIntentListener, cancellationToken: cancellationToken);
+
+        await _desktopAgent.StartAsync(cancellationToken);
+
+        if (_options.ChannelId != null)
```

Comment:
Can we move this logic to `Fdc3DesktopAgent`'s `StartAsync` method instead? This behavior is independent of the messaging layer. `HandleAddUserChannel` can be completely removed since it is only called by tests (those should test `Fdc3DesktopAgent` instead) and this method which discards the return value anyway.

Reply:
So UserChannel is currently implemented with the MessageRouter layer, so I can't pass the IMessageRouter client to the StartAsync method of the Fdc3DesktopAgent class without losing the messaging layer independency. 
It's okay to use the Fdc3DesktopAgentOptions within the Fdc3DesktopAgent class but I can't instantiate a UserChannel class without refactoring that class. (previously I was told to avoid that in this PR.)